vamb/h_loss.py

```diff
@@ -0,0 +1,820 @@
+"""Hierarchical loss for the labels suggested in https://arxiv.org/abs/2210.10929"""
+
+
+__cmd_doc__ = """Hierarchical loss for the labels"""
+
+import torch as _torch
+from torch import nn as _nn
+from torch.utils.data.dataset import TensorDataset as _TensorDataset
+import torch.nn.functional as F
+from typing import Optional, IO, Union
+from pathlib import Path
+from torch.utils.data.dataset import TensorDataset as _TensorDataset
+from torch.utils.data import DataLoader as _DataLoader
+from torch.optim import Adam as _Adam
+from torch import Tensor
+_torch.manual_seed(0)
+from collections import namedtuple
+
+import networkx as nx
+
+import numpy as _np
+from functools import partial
+
+from math import log as _log
+
+import vamb.semisupervised_encode as _semisupervised_encode
+import vamb.encode as _encode
+import vamb.hlosses_fast as _hlosses_fast
+import vamb.hier as _hier
+import vamb.infer as _infer
+
+if _torch.__version__ < '0.4':
+    raise ImportError('PyTorch version must be 0.4 or newer')
```

Comment:
Probably not necessary - we keep the dependencies in the `pyproject.toml` file, so the user should not be able to install Vamb with an outdated version of pytorch. If necessary, we can bump the compat version as high as you want.